### PR TITLE
Modify the generation format of the appendix page configuration properties table

### PR DESCRIPTION
**Pull request by hui1110**
1. Combine `default` and `description` in the current [page ](https://microsoft.github.io/spring-cloud-azure/current/reference/html/appendix.html##configuration-properties)into one column.
2. Generate the Markdown file of the properties configuration table.

### CODE REVIEW

**Comment by stliu** [comment]
I don't think this is the right design, discuss with me

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+		}
+
+		private String paddingWithChar(String printableStr, char c, int maxLength) {
+			String trimmedString = printableStr.trim();
+			return printableStr + String.join("", Collections.nCopies(maxLength - trimmedString.length(), c + ""));
+		}
+
+		private String generateAnchorName(String outputFile) {
+			outputFile = outputFile.substring(outputFile.lastIndexOf("/") + 26, outputFile.lastIndexOf("."))
+					.replace("-", " ");
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:162]: replace('-', ' ')

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+			if (outputFile.equals("all")) {
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:163]: "all".equals(outputFile)

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+				return "List of configuration";
+			}
+			String result = " ";
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:166]: StringBuilder

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+			String[] value = outputFile.split(" ");
+			for (String s : value) {
+				if (s.equals("db") | s.equals("jms") | s.equals("b2c")) {
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:169]: "db".equals(s)

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+					result = result.concat(s.toUpperCase() + " ");
+				}
+				else {
+					result = result.concat(s.substring(0, 1).toUpperCase().concat(s.substring(1)) + " ");
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:173]: extract this method to capitalize(String s)

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 6]
```diff
@@ -0,0 +1,255 @@
...
+
+        /**
+         * Get configuration properties name and descriptions.
+         */
+        protected Resource[] getResources() throws IOException {
+            return new PathMatchingResourcePatternResolver()
+                    .getResources("classpath*:/META-INF/spring-configuration-metadata.json");
+        }
+
+        protected String getCurrentDate() {
```
Comment (saragluna) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:148]: we can change the scope to private

**Review by saragluna — APPROVED** [review]

File: checkstyle-suppressions.xml [hunk 7]
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0"?>
+<!DOCTYPE suppressions PUBLIC
+        "-//Puppy Crawl//DTD Suppressions 1.1//EN"
+        "https://www.puppycrawl.com/dtds/suppressions_1_1.dtd">
+
+<suppressions>
+    <suppress files=".*.java"
+              checks="RegexpSinglelineJava"/>
```
Comment (saragluna) [inline checkstyle-suppressions.xml:8]: Should we put `files` and `checks` on the same line?

File: checkstyle-suppressions.xml [hunk 8]
```diff
@@ -0,0 +1,9 @@
...
+</suppressions>
```
Comment (saragluna) [inline checkstyle-suppressions.xml:9]: Please add one more empty line at the end of file.

**Review by saragluna — CHANGES_REQUESTED** [review]

File: checkstyle-suppressions.xml (hunk 8 above)
Comment (hui1110) [reply checkstyle-suppressions.xml]: updated

**Review by saragluna — APPROVED** [review]
LGTM.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 9]
```diff
@@ -0,0 +1,255 @@
...
+            if ("db".equals(letters) | "jms".equals(letters) | "b2c".equals(letters)) {
+                return letters.toUpperCase();
+            }
+            else {
+                return letters.substring(0, 1).toUpperCase().concat(letters.substring(1));
+            }
+        }
+
+        private String generateAnchorName(String outputFile) {
+            outputFile = outputFile.substring(outputFile.lastIndexOf("/") + 26, outputFile.lastIndexOf("."))
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:170]: Don't use `26`, use `"_configuration-properties-".length()` instead.
This requires that `outputFile` should have prefix `_configuration-properties-`, add this requirement into java doc of `Generator#generate`

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 10]
```diff
@@ -0,0 +1,217 @@
...
+import java.util.stream.Collectors;
+
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import org.springframework.core.io.Resource;
+import org.springframework.core.io.support.PathMatchingResourcePatternResolver;
+import org.springframework.util.StreamUtils;
+
+/**
+ * Generate Markdown files.
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:44]: Add more description about How to use this `Main`.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 11]
```diff
@@ -0,0 +1,217 @@
...
+		File parent = new File(outputFile).getParentFile();
+		if (!parent.exists()) {
+			System.out.println(
+					"No parent directory [" + parent + "] found. Will not generate the configuration properties file");
+			return;
+		}
+		new Generator().generate(outputFile, inclusionPattern, getCurrentDate());
+	}
+
+	static String getCurrentDate() {
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:63]: nit: `getCurrentDateString`

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 12]
```diff
@@ -0,0 +1,217 @@
...
+
+	/**
+	 * Generate Markdown files based on prefix and filename.
+	 */
+	static class Generator {
+
+		/**
+		 * Generate files prefixed with _configuration-properties.
+		 * @param outputFile the output file
+		 * @param inclusionPattern the inclusion pattern
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:78]: Need description of `date`.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 13]
```diff
@@ -0,0 +1,217 @@
...
+/**
+ * Generate Markdown files.
+ */
+public final class Main {
+
+	private Main() {
+	}
+
+	public static void main(String... args) {
+		String outputFile = args[0].replace("adoc", "md").replace("_", "");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:52]: This logic should put in the `Main`'s java doc.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+		String inclusionPattern = args.length > 1 ? args[1] : ".*";
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:53]: This logic should put in the `Main`'s java doc.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java (hunk 11 above)
Comment (chenrujun) [reply spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:63]: This logic should put in the `Main`'s java doc.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 15]
```diff
@@ -0,0 +1,217 @@
...
+		 * Generate files prefixed with _configuration-properties.
+		 * @param outputFile the output file
+		 * @param inclusionPattern the inclusion pattern
+		 */
+		void generate(String outputFile, String inclusionPattern, String date) {
+			try {
+				System.out.println("Parsing all configuration metadata");
+				List<Resource> resources = getSpringConfigurationMetadataJsonFilesInClasspath();
+				Map<String, Object> descriptions = new HashMap<>();
+				int count = 0;
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:85]: How about deleting `count` and use `resources.size()` instead?

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 16]
```diff
@@ -0,0 +1,217 @@
...
+			String[] value = outputFile.split(" ");
+			StringBuilder result = new StringBuilder(" ");
+			for (String s : value) {
+				result.append(capitalize(s)).append(" ");
+			}
+			return result.toString().trim();
+		}
+
+		private void generatePropertiesFile(String outputFile, Map<String, Object> descriptions, String date) {
+			int nameColumnWidth = Objects
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:171]:
```java
int nameColumnWidth = descriptions.keySet().stream()
					.map(String::trim)
					.mapToInt(String::length)
					.max()
					.orElse(0);
```

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 17]
```diff
@@ -0,0 +1,217 @@
...
+					.requireNonNull(
+							descriptions.keySet().stream().max(Comparator.comparingInt(String::length)).orElse(null))
+					.trim().length();
+			int descriptionColumnWidth = descriptions.values().stream().map(Object::toString)
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:175]: Same here.

Comment (chenrujun) [reply spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:175]: Furthermore, define method like this:
```java
int getMaxLength(Stream<String> stringStream) {
	return stringStream
			.map(String::trim)
			.mapToInt(String::length)
			.max()
			.orElse(0);
}
```

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 18]
```diff
@@ -0,0 +1,217 @@
...
+		}
+
+		private String paddingWithChar(String string, char c, int resultLength) {
+			if ('-' == c) {
+				return String.format("%1$-" + resultLength + "s", string.trim()).replace(' ', c);
+			}
+			else {
+				return String.format("%1$-" + resultLength + "s", string.trim());
+			}
+		}
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:145]:
```java
private String paddingWithChar(String string, char c, int resultLength) {
	String result = String.format("%1$-" + resultLength + "s", string.trim());
	if (' ' != c) {
		return result.replace(' ', c);
	}
	return result;
}
```

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 19]
```diff
@@ -0,0 +1,217 @@
...
+
+		/**
+		 * Get all "spring-configuration-metadata.json" files in classpath.
+		 */
+		protected List<Resource> getSpringConfigurationMetadataJsonFilesInClasspath() throws IOException {
+			Resource[] resources = new PathMatchingResourcePatternResolver()
+					.getResources("classpath*:/META-INF/spring-configuration-metadata.json");
+			return Arrays.stream(resources).filter(resource -> {
+				try {
+					return resource.getURL().toString().contains("azure");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:129]: Did you check this? Does this condition can filter out all target files and exclude all unnecessary files?

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 20]
```diff
@@ -0,0 +1,217 @@
...
+			if ("db".equals(letters) | "jms".equals(letters) | "b2c".equals(letters)) {
+				return letters.toUpperCase();
+			}
+			else {
+				return letters.substring(0, 1).toUpperCase().concat(letters.substring(1));
+			}
+		}
+
+		private String generateAnchorName(String outputFile) {
+			outputFile = outputFile.substring(outputFile.lastIndexOf("/") + "_configuration-properties-".length(),
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:157]: Add this logic's description into the java doc of `Main`.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 21]
```diff
@@ -0,0 +1,110 @@
...
+import static org.assertj.core.api.BDDAssertions.then;
+
+class GeneratorTests {
+
+	private static final String DATE = "MM/dd/yyyy";
+
+	private static final String INCLUSION_PATTERN = "spring.cloud.azure.*";
+
+	@Test
+	void notCreateFileNoPropertiesFound() throws URISyntaxException {
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:44]: `notCreateFileWenNoPropertiesFound`

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 22]
```diff
@@ -0,0 +1,110 @@
...
+			}
+		};
+		File file = getOutputFilePath();
+		String outputFile = file.getPath().replaceAll("\\\\", "/");
+		generator.generate(outputFile, INCLUSION_PATTERN, DATE);
+		then(file).doesNotExist();
+	}
+
+	@Test
+	void CreateFilePropertiesFound() throws URISyntaxException {
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:58]: `createFileWhenPropertiesFound`

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 23]
```diff
@@ -0,0 +1,110 @@
...
+		Main.Generator generator = new Main.Generator() {
+			@Override
+			protected List<Resource> getSpringConfigurationMetadataJsonFilesInClasspath() throws IOException {
+				Resource[] resources = new PathMatchingResourcePatternResolver()
+						.getResources("/with-azure-in-name.json");
+				return Arrays.stream(resources).filter(resource -> {
+					try {
+						return resource.getURL().toString().contains("azure");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:66]: Why need this here?

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java [hunk 24]
```diff
@@ -0,0 +1,36 @@
...
+import java.io.File;
+
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.BDDAssertions.then;
+
+class MainTests {
+
+	@Test
+	void should_do_nothing_when_parent_dir_is_not_found() {
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java:28]: Rename this method.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 25]
```diff
@@ -0,0 +1,110 @@
...
+		catch (IOException e) {
+			e.printStackTrace();
+		}
+		return result;
+	}
+
+	private File getOutputFilePath() throws URISyntaxException {
+		URL root = GeneratorTests.class.getResource(".");
+		assert root != null;
+		return new File(root.toURI().toString().substring(6), "configuration-properties-output.md");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:107]: Could you please find a better way to read this file? For example: https://stackoverflow.com/questions/1464291/how-to-really-read-text-file-from-classpath-in-java

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 26]
```diff
@@ -0,0 +1,110 @@
...
+			BufferedInputStream inFile2 = new BufferedInputStream(new FileInputStream(file2));
+			if (inFile1.available() == inFile2.available()) {
+				while (inFile1.read() != -1 && inFile2.read() != -1) {
+					if (inFile1.read() == inFile2.read()) {
+						result = true;
+					}
+				}
+			}
+			inFile1.close();
+			inFile2.close();
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:96]: Can we use `tryResourceClose` here?

Refs: https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java (hunk 19 above)
Comment (hui1110) [reply spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:129]: yes, it can

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java (hunk 22 above)
Comment (chenrujun) [reply spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:58]: I unresolved this conversation because this method still needs to update.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java (hunk 23 above)
Comment (chenrujun) [reply spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:66]: I unresolved this conversation because this question is not answered yet.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java (hunk 25 above)
Comment (chenrujun) [reply spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:107]: `substring(6)` contains magic number. Please use another way.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/resources/with-azure-in-name.json [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+    },
+    {
+      "defaultValue": "true",
+      "name": "spring.cloud.azure.second-property",
+      "description": "Second Description",
+      "type": "java.lang.Boolean"
+    },
+    {
+      "defaultValue": "true",
+      "name": "spring.cloud.third-property",
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/resources/with-azure-in-name.json:17]: Good test case.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/resources/with-azure-in-name.json [hunk 28]
```diff
@@ -0,0 +1,22 @@
+{
+  "properties": [
+    {
+      "defaultValue": "false",
+      "name": "spring.cloud.azure.first-property",
+      "description": "First Description",
+      "type": "java.lang.Boolean"
+    },
+    {
+      "defaultValue": "true",
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/resources/with-azure-in-name.json:10]: Could you please add another entry that does not contains a `defaultVaule`?

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+ */
+public final class Main {
+
+	private Main() {
+	}
+
+	/**
+	 * When the parent dictionary is found, change the suffix of the output file to md and
+	 * remove the "_" from the filename, generate configuration properties markdown files.
+	 * @param args arg[0] the output file, arg[1] the properties inclusion pattern
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:55]:
```
This method is used to ...

Arguments:
1. args[0]: output file name. The file name must start with... And in actual outputted file name, ".adoc" will be replaced by ".md", "_" will be replace by " ". For example: ...  This is required because ...
2. args[1]: 
```

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 30]
```diff
@@ -0,0 +1,256 @@
...
+	}
+
+	/**
+	 * Generate Markdown files based on properties and filename.
+	 */
+	static class Generator {
+
+		/**
+		 * Generate files prefixed with configuration-properties.
+		 * @param outputFile the output file
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:87]: The file name must ...

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 31]
```diff
@@ -0,0 +1,256 @@
...
+				}
+				generatePropertiesFile(outputFile, descriptions, date);
+			}
+			catch (IOException e) {
+				throw new IllegalStateException(e);
+			}
+		}
+
+		/**
+		 * Get all "spring-configuration-metadata.json" files in classpath.
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:131]: Not `all`, just `Spring on Azure related`.

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 32]
```diff
@@ -0,0 +1,256 @@
...
+				}
+				return false;
+			}).collect(Collectors.toList());
+		}
+
+		/**
+		 * Padding the string to the required length.
+		 * @param string the string to padding
+		 * @param c the padding content
+		 * @param resultLength the length of padding
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:154]: `The length of padding` is confusing, does it mean the count of `@param c`?

File: spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java [hunk 33]
```diff
@@ -0,0 +1,256 @@
...
+			if ("db".equals(letters) | "jms".equals(letters) | "b2c".equals(letters)) {
+				return letters.toUpperCase();
+			}
+			else {
+				return letters.substring(0, 1).toUpperCase().concat(letters.substring(1));
+			}
+		}
+
+		/**
+		 * Generate anchor name for markdown files.
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/main/java/com/azure/spring/cloud/internal/Main.java:180]: nit: `anchor` -> `title`.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 34]
```diff
@@ -0,0 +1,110 @@
...
+		Main.Generator generator = new Main.Generator() {
+			@Override
+			protected List<Resource> getSpringConfigurationMetadataJsonFilesInClasspath() throws IOException {
+				Resource[] resources = new PathMatchingResourcePatternResolver()
+						.getResources("/with-azure-in-name.json");
+				return Arrays.asList(resources);
+			}
+		};
+		File file = getOutputFilePath();
+		String outputFile = file.getPath().replaceAll("\\\\", "/");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:75]: Why is this necessary? Can we use this: https://www.baeldung.com/java-file-vs-file-path-separator

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 35]
```diff
@@ -0,0 +1,110 @@
...
+
+	private Boolean compareFile(String file2) {
+		boolean result = false;
+		try (BufferedInputStream inFile1 = new BufferedInputStream(
+				new FileInputStream("src/test/resources/configuration-properties-output.md"));
+				BufferedInputStream inFile2 = new BufferedInputStream(new FileInputStream(file2))) {
+			if (inFile1.available() == inFile2.available()) {
+				while (inFile1.read() != -1 && inFile2.read() != -1) {
+					if (inFile1.read() == inFile2.read()) {
+						result = true;
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:89]: This is not right. 

Better ways: https://stackoverflow.com/questions/27379059/determine-if-two-files-store-the-same-content

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 36]
```diff
@@ -0,0 +1,110 @@
...
+		catch (IOException e) {
+			e.printStackTrace();
+		}
+		return result;
+	}
+
+	private File getOutputFilePath() {
+		URL root = this.getClass().getResource(".");
+		if (root == null) {
+			throw new IllegalArgumentException("The output file path is not found!");
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:103]: When will this exception happen?

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java [hunk 37]
```diff
@@ -0,0 +1,36 @@
...
+import org.junit.jupiter.api.Test;
+
+import static org.assertj.core.api.BDDAssertions.then;
+
+class MainTests {
+
+	@Test
+	void notCreateFileWhenNoParentDirectoryFound() {
+		File nonExistantFile = new File("/this/file/does/not/exist");
+
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java:30]: Delete useless empty line.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java [hunk 38]
```diff
@@ -0,0 +1,36 @@
...
+		Main.main(nonExistantFile.getAbsolutePath());
+
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java:32]: Same here.

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java [hunk 39]
```diff
@@ -0,0 +1,36 @@
...
+		then(nonExistantFile).doesNotExist();
+	}
+
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/MainTests.java:35]: Do we need another test for file exists?

**Review by chenrujun — CHANGES_REQUESTED** [review]

File: spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java [hunk 40]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	private void compareFile(String file2) {
+		try (BufferedInputStream inFile1 = new BufferedInputStream(
+				new FileInputStream("src/test/resources/configuration-properties-output.md"));
+				BufferedInputStream inFile2 = new BufferedInputStream(new FileInputStream(file2))) {
+			Assertions.assertThat(inFile1).hasSameContentAs(inFile2);
+		}
+		catch (IOException e) {
+			e.printStackTrace();
```
Comment (chenrujun) [inline spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:89]: Test will pass if this exception happens.

Comment (hui1110) [reply spring-cloud-azure-markdown-appdendix-generator/src/test/java/com/azure/spring/cloud/internal/GeneratorTests.java:89]: updated

**Review by chenrujun — APPROVED** [review]
LGTM.
